frontend/javascripts/oxalis/shaders/texture_access.glsl.ts

```diff
@@ -104,56 +175,66 @@ export const getColorForCoords: ShaderModule = {
       // Will hold [highValue, lowValue];
       vec4 returnValue[2];
 
-      vec3 coords = floor(getRelativeCoords(worldPositionUVW, zoomStep));
-      vec3 relativeBucketPosition = div(coords, bucketWidth);
-      vec3 offsetInBucket = mod(coords, bucketWidth);
-
-      if (relativeBucketPosition.x > addressSpaceDimensions.x ||
-          relativeBucketPosition.y > addressSpaceDimensions.y ||
-          relativeBucketPosition.z > addressSpaceDimensions.z ||
-          relativeBucketPosition.x < 0.0 ||
-          relativeBucketPosition.y < 0.0 ||
-          relativeBucketPosition.z < 0.0) {
-        // In theory, the current magnification should always be selected
-        // so that we won't have to address data outside of the addresSpaceDimensions.
-        // Nevertheless, we explicitly guard against this situation here to avoid
-        // rendering wrong data.
-        returnValue[1] = vec4(1.0, 1.0, 0.0, 1.0);
+      if (worldPositionUVW.x < 0. || worldPositionUVW.y < 0. || worldPositionUVW.z < 0.) {
+        // should alpha be -1.0?
+        returnValue[1] = vec4(0.0, 0.0, 0.0, 0.0);
         return returnValue;
       }
 
-      float bucketIdx = linearizeVec3ToIndex(relativeBucketPosition, addressSpaceDimensions);
-
-      vec2 bucketAddressWithZoomStep = getRgbaAtIndex(
-        lookUpTexture,
-        l_texture_width,
-        bucketIdx
-      ).rg;
-
-      float bucketAddress = bucketAddressWithZoomStep.x;
-      float renderedZoomStep = bucketAddressWithZoomStep.y;
-
-      if (bucketAddress == -2.0) {
-        // The bucket is out of bounds. Render black
-        // In flight mode, it can happen that buckets were not passed to the GPU
-        // since the approximate implementation of the bucket picker missed the bucket.
-        // We simply handle this case as if the bucket was not yet loaded which means
-        // that fallback data is loaded.
-        // The downside is that data which does not exist, will be rendered gray instead of black.
-        // Issue to track progress: #3446
-        float alpha = isFlightMode() ? -1.0 : 0.0;
-        returnValue[1] = vec4(0.0, 0.0, 0.0, alpha);
-        return returnValue;
+      uint globalLayerIndex = availableLayerIndexToGlobalLayerIndex[uint(localLayerIndex)];
+      uint activeMagIdx = uint(activeMagIndices[int(globalLayerIndex)]);
+
+      float bucketAddress;
+      vec3 offsetInBucket;
+      uint renderedMagIdx;
+
+      if (!supportsPrecomputedBucketAddress) {
+        for (uint i = 0u; i < 4u; i++) {
```

Comment:
Is the `< 4` coming from the maximum allowed zoom step diff that is still rendered? If so, I'd reuse the respective constant via templating.